onnxruntime/core/providers/webgpu/buffer_manager.h

```diff
@@ -79,9 +91,39 @@ class BufferManager {
   std::vector<wgpu::Buffer> pending_staging_buffers_;
 };
 
+//
+// BufferManagerUMA needn't use staging buffers for upload and download operations. It only works with UMA GPUs.
+//
+class BufferManagerUMA : public BufferManager {
+ public:
+  explicit BufferManagerUMA(WebGpuContext& context) : BufferManager(context) {}
+  ~BufferManagerUMA() override;
+
+  void OnBufferRelease(WGPUBuffer buffer);
+
+ private:
+  void Initialize(BufferCacheMode storage_buffer_cache_mode, BufferCacheMode uniform_buffer_cache_mode,
+                  BufferCacheMode query_resolve_buffer_cache_mode) override;
+
+  WGPUBuffer Create(size_t size, wgpu::BufferUsage usage) override;
+  void Upload(void* src, WGPUBuffer dst, size_t size) override;
+  void Download(WGPUBuffer src, void* dst, size_t size) override;
+  void MemCpy(WGPUBuffer src, WGPUBuffer dst, size_t size) override;
+  void RefreshPendingBuffers() override;
+
+  std::unique_ptr<IBufferCacheManager> CreateBufferCacheManager(BufferCacheMode cache_mode) override;
+
+  // Check if the buffer was created by the class with the extended map usages.
+  bool IsUMABuffer(WGPUBuffer buffer) const;
+  // All the buffers created by the class with the extended map usages.
+  std::unordered_map<WGPUBuffer, bool> uma_buffers_;
```

Comment:
In fact, when usage is `usage & wgpu::BufferUsage::Storage`, we will use UMA buffer. So remove  `IsUMABuffer` and `uma_buffers_`, instead to use `WGPUBufferUsage wgpuBufferGetUsage(WGPUBuffer buffer) == wgpu::BufferUsage::Storage` where need to verify whether the buffer is UMA. If so, the code will be simple and clean.

Reply:
Good point. It would be simpler. Thanks!